ceremonies/src/lib.rs

```diff
@@ -384,55 +412,66 @@ impl<T: Trait> Module<T> {
             let mut reputables = Vec::with_capacity(pcount as usize);
+            let mut endorsees = Vec::with_capacity(ecount as usize);
             let mut newbies = Vec::with_capacity(pcount as usize);
 
             // TODO: upfront random permutation
             for p in 1..=pcount {
                 let participant = <ParticipantRegistry<T>>::get((cid, cindex), &p);
                 if Self::participant_reputation((cid, cindex), &participant)
                     == Reputation::UnverifiedReputable
-                    || <encointer_currencies::Module<T>>::bootstrappers(cid).contains(&participant)
                 {
                     reputables.push(participant);
+                } else if <encointer_currencies::Module<T>>::bootstrappers(cid)
+                    .contains(&participant)
+                {
+                    bootstrappers.push(participant);
+                } else if <Endorsees<T>>::contains_key((cid, cindex), &participant) {
+                    endorsees.push(participant)
                 } else {
                     newbies.push(participant);
                 }
             }
-            let mut n = reputables.len();
-            n += min(newbies.len(), n / 4);
+            let mut n = bootstrappers.len() + reputables.len();
+            n += min(newbies.len(), n / 3);
+
+            let mut all_participants = bootstrappers
+                .into_iter()
+                .chain(reputables.into_iter())
+                .chain(endorsees.into_iter())
+                .chain(newbies.into_iter());
+
             let n_meetups = n / 12 + 1;
             let mut meetups = Vec::with_capacity(n_meetups);
-            let mut meetup_n_rep = vec![0; n_meetups];
             for _i in 0..n_meetups {
                 meetups.push(Vec::with_capacity(12))
             }
-            // first, evenly assign reputables to meetups
-            for (i, p) in reputables.iter().enumerate() {
+
+            // fill meetup slots one by one in this order: bootstrappers, reputables, endorsees, newbies
+            for (i, p) in all_participants.by_ref().take(n).enumerate() {
                 meetups[i % n_meetups].push(p);
-                meetup_n_rep[i % n_meetups] += 1;
             }
-            // now, distribute newbies, complying with newbie limit per meetup
-            // FIXME: stop after skipping n_meetups newbies
-            for (i, p) in newbies.iter().enumerate() {
-                let _idx = i % n_meetups;
-                if meetups[_idx].len() < meetup_n_rep[_idx] * 4 / 3 {
-                    meetups[i % n_meetups].push(p);
-                } else {
-                    debug::debug!(target: LOG, 
-                        "had to skip a newbie: {:?}", 
-                        p);
-                }
+
+            for p in all_participants {
+                debug::debug!(target: LOG, "had to skip a newbie: {:?}", p);
             }
             // purge meetups that are too small
```

Comment:
Is this still necessary with the above logic? can there be too-small meetups? We should rather cover this with unit tests I think (test all possible numbers of all types of participants from 0 to 12 each)?

Reply:
Yes, I agree we should remove that.
And yes, I believe we should test these combinations in unit test.